sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala

```diff
@@ -241,30 +244,115 @@ object Value {
         case (ContractId.V1(hash1, suffix1), ContractId.V1(hash2, suffix2)) =>
           hash1 ?|? hash2 |+| suffix1 ?|? suffix2
       }
+    }
+
+    final case class V2 private (local: Bytes, suffix: Bytes) extends ContractId with data.NoCopy {
+      override lazy val toBytes: Bytes = V2.prefix ++ local ++ suffix
+      lazy val coid: Ref.HexString = toBytes.toHexString
+      override def toString: String = s"ContractId($coid)"
+    }
+
+    object V2 {
+      private[lf] def MaxSuffixLength: Int = V1.MaxSuffixLength
+
+      val prefix: Bytes = Bytes.assertFromString("01")
+
+      private[lf] val timePrefixSize: Int = 5
+      private[lf] val discriminatorSize: Int = 7
+
+      val localSize: Int = timePrefixSize + discriminatorSize
+
+      private val suffixStart: Int = prefix.length + localSize
+
+      def build(local: Bytes, suffix: Bytes): Either[String, V2] = {
+        for {
+          _ <- Either.cond(
+            local.length == localSize,
+            (),
+            s"The local prefix has the wrong size, expected $localSize, but got ${local.length}",
+          )
+          _ <- Either.cond(
+            suffix.length <= MaxSuffixLength,
+            (),
+            s"The suffix is too long, expected at most $MaxSuffixLength bytes, but got ${suffix.length}",
+          )
+        } yield new V2(local, suffix)
+      }
+
+      def assertBuild(local: Bytes, suffix: Bytes): V2 =
+        assertRight(build(local, suffix))
+
+      /** The largest integer `i` so that the number of microseconds between [[Time.Timestamp.MinValue]]
+        * and [[Time.Timestamp.MaxValue]] divided by `i` is smaller than `2^40-1` and therefore fits into 5 bytes.
+        */
+      private[lf] val resolution: Long = 286981L
+
+      private[lf] def timePrefix(time: Time.Timestamp): Bytes = {
+        val microsRebased = time.micros - Time.Timestamp.MinValue.micros
```

Comment:
should we add a comment that we know this does not overflow ?

Reply:
or it does not matter as we work wit unsigned long ?

Reply:
Exactly. Overflow doesn't matter here. I'll add a comment.